### PR TITLE
Add build-time handler generation with annotation processor

### PR DESCRIPTION
**Pull request by jbachorik**
## Summary

This PR adds compile-time handler generation via annotation processor as an alternative to runtime bytecode generation. The feature achieves **85% allocation reduction** with **equivalent throughput**, making it ideal for high-throughput JFR analysis pipelines and memory-constrained environments.

## Key Features

- **Annotation Processor**: Scans `@JfrType` annotated interfaces and generates handler implementations during compilation
- **ServiceLoader Auto-Discovery**: Factories automatically discovered at runtime - no manual registration needed
- **Factory Pattern**: `HandlerFactory<T>` interface with thread-local caching for reduced allocations
- **Runtime Binding**: Type IDs injected when recording is opened (not bound to specific JFR files)
- **Backward Compatible**: Falls back to runtime generation if no factories found

## Performance Results

**Benchmark Summary** (from BuildTimeHandlerBenchmark on Apple Silicon, JDK 21):

| Metric | Runtime Generation | Build-Time Generation | Improvement |
|--------|-------------------|----------------------|-------------|
| **Throughput** | 0.189 ops/s | 0.187 ops/s | ~0% (equivalent) |
| **Allocation Rate** | 237.2 MB/sec | 35.5 MB/sec | **-85%** |
| **Allocation per Op** | 223.5 MB/op | 37.4 MB/op | **-83%** |
| **GC Collections** | 3 (11ms) | 0 | **Eliminates GC** |

### New Modules
- **jafar-processor**: Annotation processor module with comprehensive tests (20 test cases)

### Core API Changes
- Added `HandlerFactory<T>` interface for factory pattern
- Added `TypedJafarParser.registerFactory()` method (optional - auto-discovery via ServiceLoader)
- Enhanced `TypeSkipper` with static skip method for generated handlers
- ServiceLoader integration for automatic factory discovery

### Generated Artifacts
- Handler implementation classes
- Factory classes with thread-local caching
- `META-INF/services/io.jafar.parser.api.HandlerFactory` (ServiceLoader registration)

### Documentation
- `doc/BUILD_TIME_BENCHMARKS.md`: Comprehensive benchmark documentation
- `doc/PERFORMANCE_REPORT.md`: Detailed performance analysis and use case recommendations
- `README.md`: Updated with build-time generation feature section

### Benchmarks
- `BuildTimeHandlerBenchmark`: Throughput, allocation profiling, and cold start tests
- 11 JFR event interfaces for realistic testing

## Usage Example

```java
// 1. Define event interface with @JfrType
@JfrType("jdk.ExecutionSample")
public interface JFRExecutionSample {
    @JfrField("startTime") long startTime();
    @JfrField("sampledThread") JFRThread sampledThread();
}

// 2. Add annotation processor dependency
dependencies {
    implementation 'io.btrace:jafar-parser:X.Y.Z'
    annotationProcessor 'io.btrace:jafar-processor:X.Y.Z'
}

// 3. Parse events - factories auto-discovered!
try (TypedJafarParser parser = ctx.newTypedParser(jfrFile)) {
    // No registration needed! ServiceLoader automatically discovers factories
    parser.handle(JFRExecutionSample.class, (event, ctl) -> {
        // Handler instance reused via thread-local caching
        process(event);
    });
    parser.run();
}
```

**That's it!** The annotation processor generates factories and registers them via ServiceLoader. No manual registration required.

## Testing

- 20 comprehensive annotation processor tests (including ServiceLoader registration verification)
- Valid cases, invalid cases, and corner cases
- JMH benchmarks with GC profiling

## Related Documentation

- [BUILD_TIME_BENCHMARKS.md](doc/BUILD_TIME_BENCHMARKS.md)
- [PERFORMANCE_REPORT.md](doc/PERFORMANCE_REPORT.md)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: jfr-shell/src/main/java/io/jafar/shell/cli/CommandDispatcher.java [hunk 1]
```diff
@@ -731,10 +731,11 @@ private void cmdTypes(List<String> args) {
...
         // Convert glob pattern to regex for string matching (not filesystem paths)
-        String globRegex = search
-            .replace(".", "\\.")  // Escape dots
-            .replace("*", ".*")   // * matches any characters
-            .replace("?", ".");   // ? matches single character
+        String globRegex =
+            search
+                .replace(".", "\\.") // Escape dots
+                .replace("*", ".*") // * matches any characters
+                .replace("?", "."); // ? matches single character
```
[the remark this answers is not in the document]
Comment (jbachorik) [reply jfr-shell/src/main/java/io/jafar/shell/cli/CommandDispatcher.java]: This was done by spotless task. Let's just keep it.

File: jafar-processor/src/main/java/io/jafar/processor/JfrTypeProcessor.java [hunk 2]
```diff
@@ -0,0 +1,601 @@
...
+      field.isArray = true;
+      if (componentType.getKind().isPrimitive()) {
+        field.needsConstantPool = false;
+        field.javaType = type.toString();
+      } else {
+        // Array of complex types - needs constant pool
+        field.needsConstantPool = true;
+        field.cpTypeName = getJfrTypeName(componentType);
+        field.javaType = "long"; // Store CP ref
+      }
```
[the remark this answers is not in the document]
Comment (jbachorik) [reply jafar-processor/src/main/java/io/jafar/processor/JfrTypeProcessor.java]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/btraceio/jafar/pull/33#discussion_r2647271076)
